core-contracts/Staking/src/main/java/finance/omm/score/staking/StakingImpl.java

```diff
@@ -433,10 +465,49 @@ private void checkForIscore() {
         BigInteger iscoreGenerated = (BigInteger) iscoreDetails.get("estimatedICX");
         if (iscoreGenerated.compareTo(BigInteger.ZERO) > 0) {
             Context.call(SYSTEM_SCORE_ADDRESS, "claimIScore");
+
+            List<Address> topPrep = getTopPreps();
+            List<Address> validPreps = getValidPreps();
+            int size = validPreps.size();
+            for (int i = 0; i < size; i++) {
+                this.validPreps.removeLast();
+            }
+            for (Address prep: topPrep) {
+                if (checkPrepStatus(prep)){
+                    this.validPreps.add(prep);
+                }
+            }
             IscoreClaimed(BigInteger.valueOf(Context.getBlockHeight()), iscoreGenerated);
         }
     }
 
+    private boolean checkPrepStatus(Address prepAddr){
+        Map<String, Object> prepDict = (Map<String, Object>) Context.call(SYSTEM_SCORE_ADDRESS, "getPRep", prepAddr);
+        BigInteger jailFlags = (BigInteger) prepDict.get("jailFlags");
+        BigInteger commissionRate = (BigInteger) prepDict.get("commissionRate");
+
+        return checkCriteria(jailFlags,commissionRate);
+    }
+
+    private boolean checkCriteria(BigInteger jailFlags, BigInteger commissionRate){
+        boolean notJailed = true;
+
+        if (jailFlags == null){
+            notJailed = true;
+        }else if (!jailFlags.equals(BigInteger.ZERO)){
+            notJailed = false;
+        }
+
+        boolean validRate = true;
+        if (commissionRate == null){
+            validRate = true;
+        }  else if (commissionRate.compareTo(getCommissionRate()) > 0) {
+            validRate = false;
+        }
+
+        return notJailed && validRate;
+    }
```

Comment:
Could avoid double negation here for readability
notJailed to jailed and then use !jailed if negation i needed

Reply:
I have made the changes [here](https://github.com/ommfinance/openmoneymarket-contracts/commit/f0e2dbb60a546e5705270d4b23656cbcb36053df)